sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java

```diff
@@ -151,72 +156,70 @@ protected String getLoggerTag() {
         final DefaultAppCenterFuture<Document<T>> result = new DefaultAppCenterFuture<>();
-
-        TokenExchange.getDbToken(
+        getTokenAndCallCosmosDbApi(
                 partition,
-                createHttpClient(mContext),
-                mApiUrl,
-                mAppSecret,
+                result,
                 new TokenExchange.TokenExchangeServiceCallback() {
                     @Override
                     public void callCosmosDb(final TokenResult tokenResult) {
-                        // https://docs.microsoft.com/en-us/rest/api/cosmos-db/get-a-document
-                        CosmosDb.callCosmosDb(
-                                tokenResult,
-                                documentId,
-                                createHttpClient(mContext),
-                                METHOD_GET,
-                                "",
-                                new ServiceCallback() {
-
-                                    @Override
-                                    public void onCallSucceeded(final String payload, Map<String, String> headers) {
-                                        result.complete(new Document<T>(payload, tokenResult.partition(), documentId));
-                                    }
-
-                                    @Override
-                                    public void onCallFailed(Exception e) {
-                                        handleApiCallFailure(e);
-                                        result.complete(new Document<T>(e));
-                                    }
-                                });
-
-                        // TODO: do we need to call `complete` here?
+                        callCosmosDbReadApi(tokenResult, documentId, result);
                     }
 
                     @Override
                     public void completeFuture(Exception e) {
-                        result.complete(new Document<T>(e));
-                    }
-                });
-
+                        completeFutureAndRemovePendingCall(e, result);
+                    }});
         return result;
     }
 
-    private TokenResult parseTokenResult(String payload) {
-        TokensResponse tokensResponse = gson.fromJson(payload, TokensResponse.class);
-        return tokensResponse.tokens().get(0);
+    private synchronized <T> void callCosmosDbReadApi(final TokenResult tokenResult, final String documentId, final DefaultAppCenterFuture<Document<T>> result) {
+        // https://docs.microsoft.com/en-us/rest/api/cosmos-db/get-a-document
+        ServiceCall cosmosDbCall = CosmosDb.callCosmosDbApi(
+                tokenResult,
+                documentId,
+                mHttpClient,
+                METHOD_GET,
+                "",
+                new ServiceCallback() {
+                    @Override
+                    public void onCallSucceeded(final String payload, Map<String, String> headers) {
```

Comment:
final not needed here, we use final on local variables or parameters only when needed.
While member/static variables must have final if they can be.